[Lens] Add search provider for global search

flash1293 opened the pull request:
Closes https://github.com/elastic/kibana/issues/77218

This PR adds a Lens entry to the new global search (ignore yellow background, that's a custom stylesheet):
<img width="630" alt="Screenshot 2020-09-15 at 10 44 17" src="https://user-images.githubusercontent.com/1508364/93188893-d499b700-f741-11ea-9aff-0034f8a5934a.png">

This is currently necessary because all apps not shown in the nav bar are not considered by the default application provider. This PR uses the same matching logic as the application provider to add a Lens entry (https://github.com/elastic/kibana/blob/c844187ee98c626c2cf1520d43ff889068d80ed9/x-pack/plugins/global_search_providers/public/providers/application.ts)

#72680 aims to add more entries to the search by considering "features" of individual apps - once that's implemented, the workaround introduced by this PR can be removed.

## Considerations

We could use the Lens visualization icon as well, but I didn't do so yet because all other "Kibana" items are using the Kibana logo (ignore yellow background, that's a custom stylesheet):
<img width="652" alt="Screenshot 2020-09-15 at 10 43 51" src="https://user-images.githubusercontent.com/1508364/93189370-5ab5fd80-f742-11ea-9d28-eed85c9775c6.png">

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74395/)
* Commit: aab4a08cdfde1f96f36cb147820b8d8b6c4b9ab2

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/aab4a08cdfde1f96f36cb147820b8d8b6c4b9ab2) | diff | [baseline](https://github.com/elastic/kibana/commit/b2ff5b43697125170f678f1c8b9afaab70111d86) |
| --- | --- | --- | --- |
| lens | 435 | +2 | 433 |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/aab4a08cdfde1f96f36cb147820b8d8b6c4b9ab2) | diff | [baseline](https://github.com/elastic/kibana/commit/b2ff5b43697125170f678f1c8b9afaab70111d86) |
| --- | --- | --- | --- |
| lens | 1.0MB | +2.2KB | 1.0MB |

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74395/","number":"74395","commit":"aab4a08cdfde1f96f36cb147820b8d8b6c4b9ab2"}]}
PIPELINE-->

elasticmachine commented:
Pinging @elastic/kibana-app (Team:KibanaApp)

elasticmachine commented:
Pinging @elastic/kibana-core-ui (Team:Core UI)

flash1293 commented:
cc @AlonaNadler

ryankeairns commented:
As Joe notes, all apps are currently using their parent solution icon. Further, saved objects are not using any icons. This was by design as we (organizationally) are moving away from the product-level logos/icons. That is not to say there aren't contrary positions, but that should likely be discussed at a broader level for the sake of consistency.

All that to say, I am a +1 for using the Kibana (solution) logo at this time.

mbondyra reviewed (approved):
Tested on FF, works good!

myasonik commented:
Congrats on being the first custom provider!!! 🎉 

@pgayvallet could you review to make sure the provider stuff is as you expected it to be used?

AlonaNadler commented:
This is awesome @flash1293 !
Few questions: 
* can we add a description next to Lens `build charts and graphs using drag and drop`?
* can we show Lens in the suggestions when users are searching for `visualize`, `chart`?

alexfrancoeur commented:
Well that was fast! Exciting to see a new provider in the mix already. As mentioned in the description, ideally we merge this into the future feature results provider (on deck). 

"Lens Visualizations" sounds to me like it will show me a filtered list of visualizations created from Lens. I understand _why_ we want visualizations in the title, but am wondering if this is the right approach. If I want to create a visualization, will the majority of users search for Lens? If I search for "visualization", would I choose Lens over Visualize? I've been meaning to open an issue to track "actions" from search. You can imagine "Create new case", "Create new alert", etc. I wonder if this is a good first candidate. "Create new visualization" or "Create new visualization from Lens" could be a good fit here if we want to start leading with Lens. I realize it's not the default experience today, but it is an opportunity to start driving more traffic to Lens with a specific task in mind.

We'll also need to make sure we have the appropriate metadata for Lens (https://github.com/elastic/kibana/issues/77114). Would the subtitle just be Kibana here? Feels like that's the right approach, but just making sure we start thinking this way with the introduction of a new provider.

ryankeairns commented:
> ...I've been meaning to open an issue to track "actions" from search. You can imagine "Create new case", "Create new alert", etc. I wonder if this is a good first candidate. "Create new visualization" or "Create new visualization from Lens" could be a good fit here if we want to start leading with Lens. I realize it's not the default experience today, but it is an opportunity to start driving more traffic to Lens with a specific task in mind.

I believe this issue covers the "actions" feature: https://github.com/elastic/kibana/issues/15019 

> We'll also need to make sure we have the appropriate metadata for Lens (#77114). Would the subtitle just be Kibana here? Feels like that's the right approach, but just making sure we start thinking this way with the introduction of a new provider.

For the metadata (i.e. subtitle text), we're heading toward using the `category` value which, for Lens, should probably be Kibana, as suggested.

------

> can we add a description next to Lens `build charts and graphs using drag and drop`?

I'm hesitant to start putting promotional text in there and would like for us to have a separate discussion about what should/should not go into that metadata field. 

> can we show Lens in the suggestions when users are searching for visualize, chart?

I'll create a separate issue for this request as it's something the search feature should provide as we would want to manage keyword ownership in a centralized location.

flash1293 commented:
> can we add a description next to Lens build charts and graphs using drag and drop?

I'm not sure, can we @myasonik ?

> can we show Lens in the suggestions when users are searching for visualize, chart

Yes, but I'm not sure whether it's confusing if the search term is not showing up in the result. @alexfrancoeur not sure how this was expected to work.

> "Create new visualization" or "Create new visualization from Lens" could be a good fit here if we want to start leading with Lens

This makes sense to me, it's also how the Visual Studio Code command palette works:
<img width="653" alt="Screenshot 2020-09-15 at 17 49 20" src="https://user-images.githubusercontent.com/1508364/93233842-d4b6a880-f77b-11ea-95cb-d94af612f9e5.png">

The action would also show up when typing "Create new"

ryankeairns commented:
@flash1293 see my reply that came in just seconds before yours :) 

My personal preference would be to progress this PR as-is and take the other requests to a separate discussion as they have global implications.

flash1293 commented:
@ryankeairns we had a little bit of a race condition here.

> I'm hesitant to start putting promotional text in there and would like for us to have a separate discussion about what should/should not go into that metadata field.

OK, I will exclude that for now

> > can we show Lens in the suggestions when users are searching for visualize, chart?

> I'll create a separate issue for this request as it's something the search feature should provide as we would want to manage keyword ownership in a centralized location.

It makes sense to create a separate issue to implement that for the general case, but for Lens specifically we could do it right now. The question is whether we should because it's not consistent with the other items

ryankeairns commented:
> > > can we show Lens in the suggestions when users are searching for visualize, chart?
> 
> > I'll create a separate issue for this request as it's something the search feature should provide as we would want to manage keyword ownership in a centralized location.
> 
> It makes sense to create a separate issue to implement that for the general case, but for Lens specifically we could do it right now. The question is whether we should because it's not consistent with the other items

I have serious reservations about doing this in this PR. Without it being centrally managed, it seems potential issues could develop if/when authors are potentially competing for the same or simliar keywords.

joshdover commented:
> I have serious reservations about doing this in this PR. Without it being centrally managed, it seems potential issues could develop if/when authors are potentially competing for the same or simliar keywords.

I share similar concerns, most from a UX perpsective. I think we want to be careful to make the search results pretty consistent between versions since users are going to build muscle memory and usage habits based on these results. So if we're going to introduce new concepts like actions or reserve keywords for specific items, we should be fairly confident that these aren't going to change significantly with each release. Fine-tuning over time is to be expected, but I don't think we want to drastically change the interactions or ordering of results on a regular basis.

All that to say that if we're going to add additional keywords for this Lens result to match on, we should be confident that this fits into the longer-term vision of how we want Global Search to respond to user input over the next few releases and that anything we match on today isn't going to be removed in the near-term.

flash1293 commented:
Got it - to summarize: We are moving out the "Actions" entries in the search, subtitles and magic matching words.

The only open question AFAICT is whether it should say "Lens Visualizations" or just "Lens". Any preference here @ryankeairns @alexfrancoeur @AlonaNadler ?

AlonaNadler commented:
`Lens Visualizations` in this point since not everyone knows what Lens is

AlonaNadler commented:
@ryankeairns regarding 

> I'm hesitant to start putting promotional text in there and would like for us to have a separate discussion about what should/should not go into that metadata field.

Its not promotional it's descriptive, many don't know what is Lens

ryankeairns commented:
> @ryankeairns regarding
> 
> > I'm hesitant to start putting promotional text in there and would like for us to have a separate discussion about what should/should not go into that metadata field.
> 
> Its not promotional it's descriptive, many don't know what is Lens

Understood, my main point is that we're currently pulling either the category (on apps) or type (on saved objects) from the existing plugin registry setup. Based upon my current understanding, there would be additional work to provide a custom description via the GS API and this also sets a new precedent for using longer descriptive text. Given those considerations, a separate issue to discuss further the approach seems an appropriate next step.

alexfrancoeur commented:
At the end of the day, we want any search related to visualizing data to be routed to Lens (chart, graph, visualize, etc.). Given that some of the earlier proposed ways require a bit more planning (https://github.com/elastic/kibana/issues/77810, https://github.com/elastic/kibana/issues/77508, https://github.com/elastic/kibana/issues/15019) it feels like we should focus on how best to word this feature entry. This is a good exercise, because we may need to address at a broader scale in the next release with https://github.com/elastic/kibana/issues/72680.

Let's spend more time kicking the tires on text. I think we all agree that Lens is not descriptive enough. I'll throw some suggestions out there, but am hoping that @gchaps might be able to help as well.

It's worth noting that for 7.10, whatever result we end up choosing with the keyword `visualize` will also return this result. So we may not want to stray too far away.

![image](https://user-images.githubusercontent.com/25181823/93527260-97325680-f906-11ea-92f0-185fec9c6045.png)

* `Visualize: Lens` - ugly, but straight to the point. This approach could eventually morph into however we end up presenting features in search
* `Create visualization with Lens` - this was my original suggestion, but would be the only task oriented result. Pushed until we address https://github.com/elastic/kibana/issues/15019.
* `Visualize with Lens` - this also feels task oriented
* `Visualize your data in Lens` - long, but similarly a task

Alternatively, we could wait until the next release and introduce the Lens results consistently with the other features and augment with searchable metadata (https://github.com/elastic/kibana/issues/77810). I only suggest this, because I'd like to avoid having too different of experiences between the next two minors. The benefits of introducing Lens in search probably outweigh the need for consistency here. If we can future proof as much as possible, it probably won't be a problem.

flash1293 commented:
I think this is great input, thanks. “Visualize: Lens” is my preference at the moment.

Maybe another option: For 7.10 - do we even need to explain what Lens is in the global search? For apps it feels more of a command palette than something a user would actually “search” for, so it could be just a shortcut to get there quickly for those who know about it already. Given the high number of apps, it’s already pretty likely a user won’t be able to make sense out of all of the entries shown when opening the search. In that case we could go with just “Lens” until we have a unified “Actions” concept.

flash1293 commented:
If you want to get a feel for it - I changed this PR to implement "Visualize: Lens" (also fixed a bug with missing permissions)
<img width="683" alt="Screenshot 2020-09-21 at 13 48 47" src="https://user-images.githubusercontent.com/1508364/93763162-3dd16c80-fc11-11ea-98d8-88737b85e07b.png">

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/75825/)
* Commit: 6035a8f4019eb605248d4df5c332228d6b931ade

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/6035a8f4019eb605248d4df5c332228d6b931ade) | diff | [baseline](https://github.com/elastic/kibana/commit/8347fd0bf3df69e5af96a86dc26d08be0a55bf5f) |
| --- | --- | --- | --- |
| lens | 449 | +2 | 447 |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/6035a8f4019eb605248d4df5c332228d6b931ade) | diff | [baseline](https://github.com/elastic/kibana/commit/8347fd0bf3df69e5af96a86dc26d08be0a55bf5f) |
| --- | --- | --- | --- |
| lens | 1.1MB | +2.7KB | 1.1MB |

### History
* :green_heart: [Build #74395](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74395/) succeeded aab4a08cdfde1f96f36cb147820b8d8b6c4b9ab2

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74395/","number":"74395","commit":"aab4a08cdfde1f96f36cb147820b8d8b6c4b9ab2"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/75825/","number":"75825","commit":"6035a8f4019eb605248d4df5c332228d6b931ade"}]}
PIPELINE-->

gchaps commented:
@alexfrancoeur, @flash1293 Here are some additional suggestions: 

* `Visualize--Lens` - might look better with an em dash than a colon (:)

* `Lens: drag, drop, visualize` - too promotional?  Could also be `Lens: create visualizations`

* `Lens visualizations` - puts Lens first

* `Visualizations with Lens` - similar to `Visualize with Lens`, but not task-oriented

AlonaNadler commented:
I think `Lens: create visualizations` is ideal
If possible lets put that first. 
We are deemphasizing visualize. So any name that highlight visualize explicitly will be there only in the short term

pgayvallet commented on x-pack/plugins/lens/public/search_provider.ts:
```diff
@@ -0,0 +1,74 @@
+      const length = Math.max(term.length, searchableTitle.length);
+      const distance = levenshtein(term, searchableTitle);
+
+      // maximum lev distance is length, we compute the match ratio (lower distance is better)
+      const ratio = Math.floor((1 - distance / length) * 100);
+      if (ratio >= 60) {
+        score = ratio;
+      }
+    }
+    if (score === 0) return EMPTY;
+    return from(
+      uiCapabilities.then(({ navLinks: { visualize: visualizeNavLink } }) =>
+        visualizeNavLink
```
NIT: the search logic is performed even when the `navLinks.visualize` capability is `false`. This could be avoided.

flash1293 commented:
Fixed the nit from Pierre and switched to `Lens: create visualizations`

kibanamachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76173/)
* Commit: 54d2d9c5df0d59ce0407734d31b7982623623db3

### Build metrics
#### @kbn/optimizer bundle module count
| id | [value](https://github.com/elastic/kibana/commit/54d2d9c5df0d59ce0407734d31b7982623623db3) | diff | [baseline](https://github.com/elastic/kibana/commit/6d819b7a1d95e8ecc50eef364ef1448544bf21da) |
| --- | --- | --- | --- |
| lens | 449 | +2 | 447 |

#### page load bundle size
| id | [value](https://github.com/elastic/kibana/commit/54d2d9c5df0d59ce0407734d31b7982623623db3) | diff | [baseline](https://github.com/elastic/kibana/commit/6d819b7a1d95e8ecc50eef364ef1448544bf21da) |
| --- | --- | --- | --- |
| lens | 1.1MB | +2.8KB | 1.1MB |

### History
* :green_heart: [Build #75825](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/75825/) succeeded 6035a8f4019eb605248d4df5c332228d6b931ade
* :green_heart: [Build #74395](https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74395/) succeeded aab4a08cdfde1f96f36cb147820b8d8b6c4b9ab2

To update your PR or re-run it, just comment with:
`@elasticmachine merge upstream`

<!--PIPELINE
{"builds":[{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/74395/","number":"74395","commit":"aab4a08cdfde1f96f36cb147820b8d8b6c4b9ab2"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/75825/","number":"75825","commit":"6035a8f4019eb605248d4df5c332228d6b931ade"},{"status":"SUCCESS","url":"https://kibana-ci.elastic.co/job/elastic+kibana+pipeline-pull-request/76173/","number":"76173","commit":"54d2d9c5df0d59ce0407734d31b7982623623db3"}]}
PIPELINE-->

alexfrancoeur commented:
I don't want to block additional discovery of lens if the team feels it's necessary to support in 7.10 vs 7.11 (both targeted). So I'm +1 on this text if the @gchaps @AlonaNadler and @ryankeairns agree.

However, if we go with `Lens: create visualizations`, we'll need to think through how this looks in the future. With a feature results provider up next (https://github.com/elastic/kibana/issues/72680), if we don't treat lens differently than other features, we'll end up with the following scenarios.

If `lens` is the input, we'd return the below results - both leading to the same view.
- `Lens`
- `Lens: create visualizations`

Very similar to how `visualiz` will work with this PR, but with different endpoints
- `Lens: create visualizations`
- `Visualize`

We'll likely have to live with these duplicate results for awhile unless we explicitly exclude lens, which isn't impossible, just an additional requirement. With that approach we likely cannot remove the lens specific registry until we have additional metadata (https://github.com/elastic/kibana/issues/77810) with the feature registry or commands (https://github.com/elastic/kibana/issues/15019). @ryankeairns and I can discuss priorities and scope.

One open question, how is the lens search registry prioritized over others? What's the hierarchy in the results?

flash1293 commented:
> One open question, how is the lens search registry prioritized over others? What's the hierarchy in the results?

The new search provider is using the same scoring algorithm than default application provider, not sure what the global search is doing on perfectly matching scores from different providers.

ryankeairns commented:
It seems we have a shared understanding that this particular change will very likely need to be adjusted/replaced in the near-ish future.Given that, I don't have any reservations about merging this as-is, especially given the importance of promoting Lens.

AlonaNadler commented:
We have a Lens first approach we need to have it in the global search, we can improve it in the future furthur but I think it is worth having it 7.10
